---
title: "SMART-Seq2 mRNA-Seq of single E7 halves from embryos injected w active/dead-Btgh - embryo filtering and clustering"
date: "August 31st, 2022"
author: 
- name: Sara Formichetti
  affiliation: EMBL Rome
  email: sara.formichetti@embl.it
output:
  prettydoc::html_pretty:
    toc: true
    toc_depth: 3
    number_sections: true
    fig_caption: yes
---

```{r Setting general options, include=FALSE}

knitr::opts_chunk$set(autodep = TRUE, cache = TRUE, echo = FALSE, warning = FALSE, message = FALSE, results = 'hide', fig.width=12, fig.height=8)

```

```{r loading needed packages}

library(ggplot2)
library(reshape2)
library(data.table)
library("DESeq2")
library(ggpubr)
library(gridExtra)
library(pheatmap)
library(ggrepel)
library(vsn)
library(RUVSeq)
library(dplyr)

```

```{r creating tx2gene and geneID2name df}

# Creating the transcriptID2geneID df; the transcriptID2geneID tsv file has been created with custom shell script tx2gene_from_encode_gtf.sh in src/sh folder
tx2gene <- fread("../../data/annotations/gencode.vM25.annotation.tx2gene.tsv", col.names = c("tx_id", "gene_id"))
# removing version
tx2gene <- tx2gene[,  lapply(.SD, gsub, pattern = "\\..*", replacement = ""), .SDcols = names(tx2gene)]
# Adding to the tx2gene DataFrame the transcript-gene pairs corresponding to OsTIR and AID
tg_tx2gene <- DataFrame(tx_id = paste0(c("NLS_EGFP_activeBtgh_NLS", "NLS_EGFP_deadBtgh_NLS"), "_transcript"), gene_id = paste0(c("NLS_EGFP_activeBtgh_NLS", "NLS_EGFP_deadBtgh_NLS"), "_gene"))
tx2gene_w_ex <- rbind(tx2gene, tg_tx2gene)

# Creating the geneID2geneName df; the geneID2geneName tsv file has been created with custom shell script geneID2name_from_encode_gtf.sh in src/sh folder
geneID2name <- fread("../../data/annotations/gencode.vM25.annotation.geneID2name.tsv", col.names = c("gene_id", "gene_name"))
# removing version
geneID2name <- geneID2name[,  lapply(.SD, gsub, pattern = "\\..*", replacement = ""), .SDcols = names(geneID2name)]
# Adding the transgenes
tg_geneID2name <- DataFrame(gene_id = paste0(c("NLS_EGFP_activeBtgh_NLS", "NLS_EGFP_deadBtgh_NLS"), "_gene"), gene_name = paste0(c("NLS_EGFP_activeBtgh_NLS", "NLS_EGFP_deadBtgh_NLS"), "_gene"))
geneID2name_w_ex <- rbind(geneID2name, tg_geneID2name)

# Merging the 2 dfs
gene_map_dt <- as.data.table(merge(geneID2name_w_ex, tx2gene_w_ex, by = "gene_id"))

```

# Design of the experiment

```{r files path}

# directory containing featureCounts outputs
dir <- "../../data/sequencing/featureCounts/STAR_default"
# vector of file names
files <- grep("_counts", sort(list.files(path = dir)), value = TRUE)
# Excluding samples that got very few number of reads
files <- files[!(grepl("A17_TR_|A4_EPI_", files))]
# adding the path
files_w_path <- file.path(dir, files)

```

```{r reading sample metadata}

# reading table containing sample names and day of collection
samples_table <- fread("../../data/sequencing/samples_table.csv")
design_df <- as.data.frame(samples_table)
# Excluding samples that got very few number of reads
design_df <- design_df[!(design_df$sample %in% c("A17_TROPHO","A4_EPI")), ]
# adding sex and tissue variables
design_df$sex <- design_df$group_or_time_point
design_df$tissue <- gsub(".*_", "", design_df$sample)
# adding condition_sex variable
design_df$condition_sex <- paste0(design_df$condition, "_", design_df$sex)
# specifying batch as a discrete variable
design_df$batch <- as.character(design_df$batch)
# reading rDNA stats
rDNA_reads <- fread("../../data/sequencing/stats/rDNA.txt", header = FALSE)
# sample name and perc of reads mapped to rDNA are in V1 and V2 columns
rDNA_reads$sample <- gsub("\\..*", "", rDNA_reads$V1)
rDNA_reads$perc_rDNA <- as.numeric(gsub("%", "", rDNA_reads$V2))
rDNA_reads <- rDNA_reads[, .(sample, perc_rDNA)]
# merging rDNA info with design df
design_df <- merge(design_df, rDNA_reads, by = "sample")
# reordering design df for visualization purposes
design_df <- design_df[order(design_df$condition, decreasing = TRUE), ]
rownames(design_df) <- design_df$sample # necessary to make dds

```

```{r design table, results='asis'}

knitr::kable(design_df, row.names = FALSE)

```

```{r reading counts}

# reading and merging all files in one data.frame by Geneid
counts_table <- as.data.frame(Reduce(function(...) merge(..., by = c("Geneid")), lapply(files_w_path, function (f) {fread(f)})))
# simplifying colnames by leaving only the sample name
colnames(counts_table) <- gsub("_TR.*", "_TROPHO", gsub("HGY5JBGXM_", "", gsub("_22s00.*", "", names(counts_table))))
# adding gene ids as rownames
rownames(counts_table) <- counts_table$Geneid
counts_table$Geneid <- NULL
# removing version from gene ids
rownames(counts_table) <- gsub("\\..*", "", rownames(counts_table))

```

# Quality filtering of embryos

```{r mt genes in metadata}

# reading mt genes
mt_genes <- fread("../../data/annotations/gencode.vM25.mt_genes.csv")
# removing version from gene id
mt_genes$gene_id <- gsub("\\..*", "", mt_genes$gene_id)

# for each sample, proportion of counts for mt genes
counts_table_qc <- as.data.table(counts_table)
counts_table_qc$gene_id <- rownames(counts_table)
mt_genes_counts <- apply(merge(counts_table_qc, mt_genes, by = "gene_id", all.x = FALSE, all.y = TRUE)[, 2:ncol(counts_table_qc)], 2, sum)
tot_counts <- apply(counts_table, 2, sum)
mt_genes_table <- data.table(sample = colnames(counts_table), mt_genes_counts = mt_genes_counts, tot_counts = tot_counts)
design_df <- merge(design_df, mt_genes_table, by = "sample")
# reordering design df for visualization purposes
design_df <- design_df[order(design_df$tissue, design_df$condition, design_df$sex, decreasing = TRUE), ]
rownames(design_df) <- design_df$sample
# reordering counts table based on sample order in design df
counts_table <- counts_table[, design_df$sample]

```

```{r quality filtering plot, include = TRUE}

ggplot(data = design_df[design_df$tissue == "EPI",], aes(x = perc_rDNA, y = mt_genes_counts/tot_counts*100, label = sample, color = condition)) +
  geom_point(size = 4) +
  geom_text_repel(size = 3) +
  xlab("perc rDNA reads") +
  ylab("perc mt genes counts") +
  ggtitle("EPI") +
  theme(title = element_text(size = 20),
        legend.text = element_text(size = 20), legend.title = element_text(size = 20),
        axis.text = element_text(size = 19),
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        panel.background = element_blank(),
        axis.line = element_line(colour = "black"))

ggplot(data = design_df[design_df$tissue == "TROPHO",], aes(x = perc_rDNA, y = mt_genes_counts/tot_counts*100, label = sample, color = condition)) +
  geom_point(size = 4) +
  geom_text_repel(size = 3) +
  xlab("perc rDNA reads") +
  ylab("perc mt genes counts") +
  ggtitle("TROPHO") +
  theme(title = element_text(size = 20),
        legend.text = element_text(size = 20), legend.title = element_text(size = 20),
        axis.text = element_text(size = 19),
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        panel.background = element_blank(),
        axis.line = element_line(colour = "black"))

```

No sample shows a particularly high mtDNA gene counts and rDNA percentage.

```{r creating DESeqDataSet}

# creating DESeqDataSet from count table
dds <- DESeqDataSetFromMatrix(countData = counts_table,
                              colData = design_df,
                              design = ~ condition)

```

```{r prefiltering}

# pre-filtering low count genes in order to speed up the following computations
keep <- rowSums(counts(dds)) >= 10
dds <- dds[keep,]

```

```{r DESeq function to compute size factors for gene counts}

# the standard differential expression analysis steps are wrapped into a single function, DESeq
dds <- DESeq(dds)

```

# Verifying sex of embryos

```{r sexing gene counts}

sexing_genes <- gene_map_dt[gene_name %in% c("Ddx3y", "Xist"), unique(gene_id)]
names(sexing_genes) <- gene_map_dt[gene_name %in% c("Ddx3y", "Xist"), unique(gene_name)]

get_norm_dds_counts <- function (my_gene_id, my_dds, my_design_df) {
  return(plotCounts(dds, gene = my_gene_id, intgroup = c("sample", "sex", "condition"), returnData = TRUE, normalized = TRUE))
}

sexing_gene_counts_list <- lapply(sexing_genes, get_norm_dds_counts, my_dds = dds, my_design_df = design_df)
sfx <- paste0(".", names(sexing_genes))
sexing_gene_counts <- sexing_gene_counts_list[[1]]
for(i in head(seq_along(sexing_gene_counts_list), n = -1L)) {
 sexing_gene_counts <- merge(sexing_gene_counts, sexing_gene_counts_list[[i+1]],
                            all = TRUE, 
                            suffixes = sfx[i:(i+1)], 
                            by = c("sample", "sex", "condition"))
}

```

```{r plot sexing counts, include = TRUE}

ggplot(sexing_gene_counts, aes(x = count.Ddx3y, y = count.Xist, colour = sex, label = sample)) +
  geom_point(size = 3) +
  geom_text_repel(size = 6) +
  theme(title = element_text(size = 20),
        legend.text = element_text(size = 20), legend.title = element_text(size = 20),
        axis.text = element_text(size = 19),
        legend.key.height = unit(1.5,"line"),
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        panel.background = element_blank(),
        axis.line = element_line(colour = "black"))

```

The sex established by cDNA genotyping is confirmed in silico.

# Testing purity of tissues

I use 3 markers of E7 epiblast and 3 markers of E7 extraembryonic ectoderm.

```{r EPI and TROPHO markers gene counts}

markers_EPI <- gene_map_dt[gene_name %in% c("Map7d3", "Pdzd4", "Uchl1"), unique(gene_id)]
markers_TROPHO <- gene_map_dt[gene_name %in% c("Gjb3", "Gm9", "Wnt6"), unique(gene_id)]
names(markers_EPI) <- gene_map_dt[gene_name %in% c("Map7d3", "Pdzd4", "Uchl1"), unique(gene_name)]
names(markers_TROPHO) <- gene_map_dt[gene_name %in% c("Gjb3", "Gm9", "Wnt6"), unique(gene_name)]

get_norm_dds_counts <- function (my_gene_id, my_dds, my_design_df) {
  return(plotCounts(dds, gene = my_gene_id, intgroup = c("sample", "condition_sex", "batch", "tissue", "sex", "condition"), returnData = TRUE, normalized = TRUE))
}

markers_EPI_gene_counts_list <- lapply(markers_EPI, get_norm_dds_counts, my_dds = dds, my_design_df = design_df)
markers_TROPHO_gene_counts_list <- lapply(markers_TROPHO, get_norm_dds_counts, my_dds = dds, my_design_df = design_df)

sfx <- paste0(".", c(names(markers_EPI), names(markers_TROPHO)))
markers_gene_counts <- markers_EPI_gene_counts_list[[1]]
for(i in head(seq_along(c(markers_EPI_gene_counts_list, markers_TROPHO_gene_counts_list)), n = -1L)) {
 markers_gene_counts <- merge(markers_gene_counts, c(markers_EPI_gene_counts_list, markers_TROPHO_gene_counts_list)[[i+1]],
                            all = TRUE, 
                            suffixes = sfx[i:(i+1)], 
                            by = c("sample", "condition_sex", "batch", "tissue", "sex", "condition"))
}
names(markers_gene_counts) <- gsub("count\\.", "", names(markers_gene_counts))
markers_gene_counts_long <- reshape2::melt(markers_gene_counts, value.name = "count", variable.name = "marker")

markers_gene_counts$EPI_avg <- rowMeans(markers_gene_counts[, names(markers_EPI)]) 
markers_gene_counts$TROPHO_avg <- rowMeans(markers_gene_counts[, names(markers_TROPHO)]) 

```

```{r plot markers counts, include = TRUE}

ggplot(markers_gene_counts_long[markers_gene_counts_long$marker %in% names(markers_EPI), ], aes(x = condition, y = count, label = sample)) +
  geom_point(aes(color = tissue, group = tissue), position = position_jitterdodge(dodge.width = 0.9, jitter.width = 0.05, seed = 1)) +
  geom_text_repel(aes(color = tissue, group = tissue), size = 2, position = position_jitterdodge(dodge.width = 0.9, jitter.width = 0.05, seed = 1)) +
  scale_y_log10() +
  facet_wrap(~marker) +
  theme(title = element_text(size = 20),
        legend.text = element_text(size = 20), legend.title = element_text(size = 20),
        axis.text = element_text(size = 19),
        axis.text.x = element_text(angle = 45, vjust = 1, hjust=1),
        legend.key.height = unit(1.5,"line"),
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        panel.background = element_blank(),
        axis.line = element_line(colour = "black"))

ggplot(markers_gene_counts_long[markers_gene_counts_long$marker %in% names(markers_TROPHO), ], aes(x = condition, y = count, label = sample)) +
  geom_point(aes(color = tissue, group = tissue), position = position_jitterdodge(dodge.width = 0.9, jitter.width = 0.05, seed = 1)) +
  geom_text_repel(aes(color = tissue, group = tissue), size = 2, position = position_jitterdodge(dodge.width = 0.9, jitter.width = 0.05, seed = 1)) +
  scale_y_log10() +
  facet_wrap(~marker) +
  theme_minimal() +
  theme(title = element_text(size = 20),
        legend.text = element_text(size = 20), legend.title = element_text(size = 20),
        axis.text = element_text(size = 19),
        axis.text.x = element_text(angle = 45, vjust = 1, hjust=1),
        legend.key.height = unit(1.5,"line"),
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        panel.background = element_blank(),
        axis.line = element_line(colour = "black"))

ggplot(markers_gene_counts, aes(x = EPI_avg, y = TROPHO_avg, label = sample)) +
  geom_point(aes(color = tissue, shape = batch)) +
  geom_text_repel(aes(color = tissue), size = 3) +
  xlab("avg counts EPI markers") +
  ylab("avg counts TROPHO markers") +
  theme(title = element_text(size = 20),
        legend.text = element_text(size = 20), legend.title = element_text(size = 20),
        axis.text = element_text(size = 19),
        legend.key.height = unit(1.5,"line"),
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        panel.background = element_blank(),
        axis.line = element_line(colour = "black"))

```

I exclude samples A22_TROPHO because outlier for the expression of both sets od markers and sample A19_TROPHO because more contaminated by EPI tissue.

```{r excluding unpure samples}

design_df <- design_df[!(design_df$sample %in% c("A22_TROPHO", "A19_TROPHO")), ]
rownames(design_df) <- design_df$sample
counts_table <- counts_table[, rownames(design_df)]

```

```{r recreating DESeqDataSet}

# removing previous dds
rm(dds)

# creating DESeqDataSet from count table
dds <- DESeqDataSetFromMatrix(countData = counts_table,
                              colData = design_df,
                              design = ~ condition)

```

```{r reprefiltering}

# pre-filtering low count genes in order to speed up the following computations
keep <- rowSums(counts(dds)) >= 10
dds <- dds[keep,]

```

```{r reDESeq function to compute size factors for gene counts}

# the standard differential expression analysis steps are wrapped into a single function, DESeq
dds <- DESeq(dds)

```

# Sample clustering

## Heatmap for quality assessment

After applying the log2(n + 1) transformation, I plot heatmaps of normalized counts for top highly expressed genes to check for eventual big sample heterogeneity present in the dataset. 

```{r pheatmap after data transformation, include = TRUE}

highest_expr_hm <- function (my_dds, my_title) {
  select <- order(rowMeans(counts(my_dds, normalized = FALSE)), decreasing = TRUE)[1:20]
  design_df_h <- as.data.frame(colData(my_dds)[,c("condition", "tissue", "sex", "batch")])
  rownames(design_df_h) <- colnames(counts(my_dds))
  my_ntd <- normTransform(my_dds)
  pheatmap(assay(my_ntd)[select,], cluster_rows = FALSE, show_rownames = TRUE, show_colnames = TRUE, cluster_cols = TRUE, annotation_col = design_df_h, main = my_title)
}

highest_expr_hm(dds, my_title = "")

```

No concerns.

## PCA

```{r defining mean threshold and var genes number threshold}

my_mean_threshold <- 10
my_ntop <- 1000

```

For this PCA:

1. I start with DESeq2-normalized data
2. I keep only genes with DESeq2-normalized mean across samples > `r my_mean_threshold`
3. I apply log2 transformation
4. I perform PCA with package prcomp using the `r my_ntop` genes with the highest variance

```{r PCA functions}

plotPCA_DESeq_norm_data <- function (my_dds, intgroup, mean_threshold, ntop, returnData = TRUE) {
    # computing row means of DESeq2-normalized counts
    rM <- rowMeans2(counts(my_dds, normalized = TRUE))
    # creating object of normTransformed data i.e. log2 transformed DESeq2-normalized data
    ntd <- normTransform(my_dds)
    # keeping only genes with rowMean > my threshold
    f_norm_counts <- assay(ntd)[rM > mean_threshold, ]
    # computing row vars of log2 transformed DESeq2-normalized counts
    rv <- rowVars(f_norm_counts)
    # selecting the first ntop genes based on var computed on line above
    select <- order(rv, decreasing = TRUE)[seq_len(min(ntop,length(rv)))]
    # pca of the selected genes
    pca <- prcomp(t(f_norm_counts[select, ]))
    percentVar <- pca$sdev^2/sum(pca$sdev^2)
    if (!all(intgroup %in% names(colData(my_dds)))) {
        stop("the argument 'intgroup' should specify columns of colData(dds)")
    }
    intgroup.df <- as.data.frame(colData(my_dds)[, intgroup, drop = FALSE])
    group <- if (length(intgroup) > 1) {
        factor(apply(intgroup.df, 1, paste, collapse = ":"))
    }
    else {
        colData(my_dds)[[intgroup]]
    }
    d <- data.frame(PC1 = pca$x[, 1], PC2 = pca$x[, 2], PC3 = pca$x[, 3], PC4 = pca$x[, 4], intgroup.df, name = colnames(my_dds))
    if (returnData) {
        attr(d, "percentVar") <- percentVar[1:4]
        return(d)
    }
}

plotPCA_DESeq_norm <- function (my_dds, intgroup, mean_threshold, ntop, PC_x, PC_y) {
  pcaData <- plotPCA_DESeq_norm_data(my_dds = my_dds, intgroup = intgroup, mean_threshold = mean_threshold, ntop = ntop, returnData = TRUE)
  percentVar <- round(100 * attr(pcaData, "percentVar"))
  pcaData[,6] <- as.character(pcaData[,6])
  par(mfrow=c(2,1))
  ggplot(pcaData, aes(pcaData[, PC_x], pcaData[, PC_y], color=pcaData[,5], label = rownames(pcaData))) +
    geom_point(size=4, aes(shape=pcaData[,6])) +
    xlab(paste0("PC",PC_x,": ",percentVar[PC_x],"% variance")) +
    ylab(paste0("PC",PC_y,": ",percentVar[PC_y],"% variance")) +
    scale_color_discrete(name = colnames(pcaData)[5]) +
    scale_shape_discrete(name = colnames(pcaData)[6]) +
    geom_text_repel(size = 3) +
    coord_fixed() +
    theme(axis.text = element_text(size = 22), axis.title = element_text(size = 22),
        title = element_text(size = 22),
        legend.text = element_text(size = 22), legend.title = element_text(size = 22),
        legend.key.height = unit(1.5,"line"),
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        panel.background = element_blank(),
        axis.line = element_line(colour = "black"))
}

```

```{r PCA plots all samples, include = TRUE}

plotPCA_DESeq_norm(my_dds = dds, intgroup = c("condition_sex", "tissue"), mean_threshold = my_mean_threshold, ntop = my_ntop, PC_x = 1, PC_y = 2)

plotPCA_DESeq_norm(my_dds = dds, intgroup = c("condition_sex", "tissue"), mean_threshold = my_mean_threshold, ntop = my_ntop, PC_x = 2, PC_y = 3)

```

As expected, the samples cluster based on the tissue on PC1, then based on sex on PC2 and PC3. I will do a separate Diff Expr analysis for each tissue. Additionally separating by sex would result in too few embryos per group (see table below) considering the variability of in vivo data.  

```{r splitting data by tissue}

# creating categorical variable batch_condition i.e. litter
design_df$litter <- paste0(design_df$condition, ".", design_df$batch)

counts_table_EPI <- counts_table[, grepl("EPI", names(counts_table))]
counts_table_TROPHO <- counts_table[, grepl("TROPHO", names(counts_table))]
design_df_EPI <- design_df[design_df$tissue == "EPI",]
design_df_TROPHO <- design_df[design_df$tissue == "TROPHO",]

```

# Number of samples after all filtering steps

The following numbers of embryos are left after all filtering steps. These will be input to the DE analysis in next Rmd.

```{r design table after filtering, results='asis'}

knitr::kable(table(design_df$tissue, design_df$condition_sex), row.names = TRUE)

```

```{r saving objects needed in next Rmd}

save(design_df, file = "Rdata/design_df.Rdata")
save(gene_map_dt, file = "Rdata/gene_map_dt.Rdata")
save(geneID2name_w_ex, file = "Rdata/geneID2name_w_ex.Rdata")
save(counts_table, file = "Rdata/counts_table.Rdata")
save(counts_table_EPI, file = "Rdata/counts_table_EPI.Rdata")
save(counts_table_TROPHO, file = "Rdata/counts_table_TROPHO.Rdata")
save(design_df_EPI, file = "Rdata/design_df_EPI.Rdata")
save(design_df_TROPHO, file = "Rdata/design_df_TROPHO.Rdata")

```

```{r sessioninfo, echo = TRUE, results = 'markup'}

sessionInfo()

```



























