---
title: "Dynamics of expression in WeiXie data (from MII oocyte to E7.5) of genes associated with PCs of DEGs PCA Btgh data with all stages"
date: "2nd November, 2023"
author: 
- name: Sara Formichetti
  affiliation: EMBL Rome
  email: sara.formichetti@embl.it
output: html_document
---

```{r Setting general options, include=FALSE}

knitr::opts_chunk$set(autodep = TRUE, cache = TRUE, echo = FALSE, warning = FALSE, message = FALSE, self.contained = TRUE)

```

```{r loading needed packages}

library("DESeq2")
library("tximport")
library(ggplot2)
library(reshape2)
library(data.table)

```

# SRA samples metadata

```{r reading sample metadata}

SRA_RunTable <- fread("../../data/sequencing/SraRunTable_selection.csv")

# counting technical replicates based on Experiment ID
SRA_RunTable[, c("tech_rep") := lapply(.SD, function(ds){seq(1, length(ds))}), .SDcols = "Consent", by = "Experiment"]
# counting biological replicates
SRA_RunTable[, c("biol_rep") := lapply(.SD, function(ds){seq(1, length(ds))}), .SDcols = "tech_rep", by = c("Developmental_Stage", "tech_rep")]
# make sample name as combination of stage, tech rep and biological rep
SRA_RunTable$sample <- paste0(gsub(" ", "_", SRA_RunTable$Developmental_Stage), ".", SRA_RunTable$biol_rep, ".", SRA_RunTable$tech_rep)
# order stage
SRA_RunTable$Developmental_Stage <- factor(SRA_RunTable$Developmental_Stage, levels = c("MII oocyte", "Zygote", "early 2-cell", "2-cell", "4-cell", "8-cell", "E3.5_ICM", "E3.5_TE", "E4.0_ICM", "Inner cell mass", "Embryonic stem cell", "E5.5_Epi", "E5.5_VE", "E6.5_Epi", "E6.5_VE", "E7.5_Ecto", "E7.5_PS", "E7.5_Meso", "E7.5_Endo"))
setorder(SRA_RunTable, Developmental_Stage, biol_rep, tech_rep)
rownames(SRA_RunTable) <- SRA_RunTable$sample # necessary to in case make dds

```

```{r design table, results='asis'}

knitr::kable(SRA_RunTable, row.names = FALSE)

```

```{r creating tx2gene and geneID2name df}

# Creating the transcriptID2geneID df; the transcriptID2geneID tsv file has been created with custom shell script tx2gene_from_encode_gtf.sh in src/sh folder
tx2gene <- fread("../../data/annotations/gencode.vM25.annotation.tx2gene.tsv", col.names = c("tx_id", "gene_id"))
# removing version
tx2gene <- tx2gene[,  lapply(.SD, gsub, pattern = "\\..*", replacement = ""), .SDcols = names(tx2gene)]

# Creating the geneID2geneName df; the geneID2geneName tsv file has been created with custom shell script geneID2name_from_encode_gtf.sh in src/sh folder
geneID2name <- fread("../../data/annotations/gencode.vM25.annotation.geneID2name.tsv", col.names = c("gene_id", "gene_name"))
# removing version
geneID2name <- geneID2name[,  lapply(.SD, gsub, pattern = "\\..*", replacement = ""), .SDcols = names(geneID2name)]

# Merging the 2 dfs
gene_map_dt <- as.data.table(merge(geneID2name, tx2gene, by = "gene_id"))

```

```{r creating the txi object}

# specifying the directory containing Salmon outputs
dir <- "../../data/sequencing/Salmon"

files <- sort(list.files(path = dir))

# making the  character vector of names of Salmon quantification's files
files_w_path <- file.path(dir, files)

# tximport gene level summarization of transcripts' quantification
txi <- tximport(files_w_path, type="salmon", tx2gene=tx2gene, ignoreTxVersion = TRUE)

# renaming columns of txi matrices object
colnames(txi$counts) <- gsub("_.*", "", files)
colnames(txi$abundance) <- gsub("_.*", "", files)
colnames(txi$length) <- gsub("_.*", "", files)

# ordering txi matrices as the samples in the SRA table
txi$counts <- txi$counts[, SRA_RunTable$Run]
txi$abundance <- txi$abundance[, SRA_RunTable$Run]
txi$length <- txi$length[, SRA_RunTable$Run]

```

```{r working on TPM matrix}

# TPM values are in the abundance matrix
TPM_dt <- data.table(txi$abundance)
# reading gene ids
TPM_dt$gene_id <- rownames(txi$abundance)
# adding gene names
TPM_dt <- merge(TPM_dt, geneID2name, by = "gene_id")
# giving sample names as column names
names(TPM_dt)[-c(1, ncol(TPM_dt))] <- SRA_RunTable$sample

```

```{r removing bad samples}

TPM_dt$`early_2-cell.2.1` <- NULL
SRA_RunTable <- SRA_RunTable[SRA_RunTable$sample != "early_2-cell.2.1"]

```

```{r loading gene lists from Btgh data}

PC1_genes <- fread("../../data/Btgh_proj/DEGs_PCA_PC1_first_50_genes.csv")
PC2_genes <- fread("../../data/Btgh_proj/DEGs_PCA_PC2_first_30_genes.csv")

```

```{r TPM lineplot function}

library(ggpubr)

gene_lineplot_preimpl <- function (gene_symbols, my_title, show_legend) {
  gene_TPM <- TPM_dt[gene_name %in% gene_symbols]
  TPM_to_plot <- melt(gene_TPM, id.vars = c("gene_name", "gene_id"), variable.name = "sample", value.name = "TPM")
  # creating stage_biolRep variable from sample name
  TPM_to_plot$stage_rep <- gsub("\\.[1-9]$", "", TPM_to_plot$sample)
  # creating stage variable from sample name
  TPM_to_plot$stage <- gsub("\\.[1-9]$", "", TPM_to_plot$stage_rep)
  # reordering the stage for plotting
  TPM_to_plot$stage <- factor(TPM_to_plot$stage, levels = c("MII_oocyte", "Zygote", "early_2-cell", "2-cell", "4-cell", "8-cell", "E3.5_ICM", "Inner_cell_mass", "E4.0_ICM", "Embryonic_stem_cell", "E5.5_Epi", "E6.5_Epi", "E7.5_Ecto", "E7.5_PS", "E7.5_Meso", "E7.5_Endo", "E3.5_TE", "E5.5_VE", "E6.5_VE"))
  # averaging TPM for tech reps
  TPM_to_plot <- TPM_to_plot[, lapply(.SD, mean), .SDcols = "TPM", by = c("stage_rep", "gene_name", "stage")]
  # averaging TPM for biol reps
  TPM_to_plot[, c("TPM_mean") := list(mean(TPM)), by = c("stage", "gene_name")]
  # separate the genes based on higher expression in ICM or TE in blastocyst
  # getting expr in ICM and TE for each gene
  blasto_TPM <- unique(TPM_to_plot[stage %in% c("E3.5_ICM", "E3.5_TE"), .(TPM_mean, gene_name, stage)])
  blasto_TPM_large <- dcast(blasto_TPM, gene_name ~ stage, value.var = "TPM_mean")
  blasto_TPM_large$diff <- blasto_TPM_large$E3.5_ICM - blasto_TPM_large$E3.5_TE
  blasto_TPM_large$blasto_tissue <- ifelse(blasto_TPM_large$diff > 0, "ICM", "TE")
  TPM_to_plot <- merge(TPM_to_plot, blasto_TPM_large[, .(gene_name, blasto_tissue)], by = "gene_name")
  TPM_to_plot_preimpl_E <- TPM_to_plot[TPM_to_plot$blasto_tissue == "ICM" & TPM_to_plot$stage %in% c("4-cell", "8-cell", "E3.5_ICM")]
  TPM_to_plot_preimpl_EE <- TPM_to_plot[TPM_to_plot$blasto_tissue == "TE" & TPM_to_plot$stage %in% c("4-cell", "8-cell", "E3.5_TE")]
  E <-  ggplot(data = TPM_to_plot_preimpl_E, aes(x = stage, y = TPM_mean)) +
        geom_point() +
        geom_line(aes(color = gene_name, group = gene_name), show.legend = show_legend) +
        ylab("TPM (log2 scale)") +
        xlab("") +
        scale_y_continuous(trans = "log2") +
        ggtitle(my_title, " > expr in ICM") +
        scale_fill_brewer(palette = "Set2") +
        theme(axis.text = element_text(size = 12), axis.title = element_text(size = 12),
          title = element_text(size = 12),
          legend.text = element_text(size = 10), legend.title = element_text(size = 10),
          legend.key.height = unit(1.5,"line"),
          panel.grid.major = element_blank(),
          panel.grid.minor = element_blank(),
          panel.background = element_blank(),
          axis.line = element_line(colour = "black"),
          axis.text.x = element_text(angle = 45, hjust = 1))
    EE <-  ggplot(data = TPM_to_plot_preimpl_EE, aes(x = stage, y = TPM_mean)) +
        geom_point() +
        geom_line(aes(color = gene_name, group = gene_name), show.legend = show_legend) +
        ylab("TPM (log2 scale)") +
        xlab("") +
        scale_y_continuous(trans = "log2") +
        ggtitle(my_title, " > expr in TE") +
        scale_fill_brewer(palette = "Set2") +
        theme(axis.text = element_text(size = 12), axis.title = element_text(size = 12),
          title = element_text(size = 12),
          legend.text = element_text(size = 10), legend.title = element_text(size = 10),
          legend.key.height = unit(1.5,"line"),
          panel.grid.major = element_blank(),
          panel.grid.minor = element_blank(),
          panel.background = element_blank(),
          axis.line = element_line(colour = "black"),
          axis.text.x = element_text(angle = 45, hjust = 1))
    ggarrange(E,EE)
}

```

```{r loadings lineplots (FIGURE), include = TRUE}

my_gene_symbols <- c(PC1_genes[coord > 0]$gene_name, PC2_genes[coord > 0]$gene_name)
gene_lineplot_preimpl(gene_symbols = my_gene_symbols, my_title = "PC1 and PC2 right", show_legend = FALSE)
my_gene_symbols <- c(PC1_genes[coord < 0]$gene_name, PC2_genes[coord < 0]$gene_name)
gene_lineplot_preimpl(gene_symbols = my_gene_symbols, my_title = "PC1 and PC2 left", show_legend = FALSE)

```

## Conclusion

Almost all DEGs mostly associated with PC1 and PC2 are following the expected trend if those PC represent developmental time i.e. upregulation if higher in PC coordinate and downregulation if lower in PC coordinate. The only 4 (out of 75) genes that go on the the wrong direction are: Mllt6, Isyma1, Dus3l and Nudt9 (found by plotting subsets of the genes and showing legend).

```{r sessioninfo, echo = TRUE, results = 'markup'}

sessionInfo()

```