---
title: "Testing maternal vs paternal chromosomal imbalance in FVB/PWD F1 hybrid blastocysts injected and nonInjected w active/dead-Btgh"
date: "September 14th, 2022"
author: 
- name: Sara Formichetti
  affiliation: EMBL Rome
  email: sara.formichetti@embl.it
output:
  prettydoc::html_pretty:
    toc: true
    toc_depth: 3
    number_sections: true
    fig_caption: yes
---

```{r Setting general options, include=FALSE}

knitr::opts_chunk$set(autodep = TRUE, cache = TRUE, echo = FALSE, warning = FALSE, message = FALSE, results = 'hide', fig.width=12, fig.height=8)

```

```{r loading needed packages}

library(ggplot2)
library(reshape2)
library(data.table)
library(ggpubr)
library(gridExtra)
library(pheatmap)
library(ggrepel)
library(dplyr)

```

```{r loading Rdata}

load("Rdata/design_df_before_aneupl_filter.Rdata")
load("Rdata/geneID2name_w_ex.Rdata")

# adding sex info
load("Rdata/design_df.Rdata")
design_df_before_aneupl_filter <- merge(design_df_before_aneupl_filter, design_df, all = TRUE)

```

```{r files path}

# directory containing featureCounts outputs
dir <- "../../data/sequencing/featureCounts/SNPsplit"
# vector of file names
files <- grep("_joined_counts.txt", sort(list.files(path = dir)), value = TRUE)
# Excluding samples that got very few number of reads or did not pass filtering based on rRNA and mtgenes in embryo_filtering_and_clustering.Rmd
files <- files[!(grepl("bC20_|bC24_|bC25_|bC28_|bA9_|bA8_|bD9_", files))]
# adding the path
files_w_path <- file.path(dir, files)

```

```{r reading counts}

# reading counts for all samples
counts_table_list <- lapply(files_w_path, function (f) {fread(f, col.names = c("gene_id", "total", "unassigned", "FVB", "PWD"))})
# list names are the sample names
names(counts_table_list) <- gsub("_.*", "", files)

```

```{r counts by chromosome}

## merging counts table w chr info
# reading table with gene ids, tx ids and gene chr
tx_gene_chr_map <- fread("../../data/annotations/gencode.vM25.annotation.tx2gene_w_chr.tsv", col.names = c("chr", "tx_id", "gene_id"))
# getting rid of tx info
gene_chr_map <- unique(tx_gene_chr_map[, .(gene_id, chr)])
# remove assembly version
gene_chr_map$gene_id <- gsub("\\..*", "", gene_chr_map$gene_id)
# adding gene names
gene_chr_map <- merge(gene_chr_map, geneID2name_w_ex, by = c("gene_id"))

count_by_chr <- function (i) {
  sample_table <- counts_table_list[[i]]
  sample_table$gene_id <- gsub("\\..*", "", sample_table$gene_id)
  sample_table <- merge(sample_table, gene_chr_map, by = "gene_id")
  sample_table <- sample_table[, lapply(.SD, sum), by = chr, .SDcols = c("total", "unassigned", "FVB", "PWD")]
  sample_table$sample <- rep(names(counts_table_list)[i], nrow(sample_table))
  return(sample_table)
}
# binding all sample tables
counts_table_by_chr_long <- do.call("rbind", lapply(1:length(counts_table_list), count_by_chr))
# adding sample metadata
counts_table_by_chr_long <- merge(counts_table_by_chr_long, design_df_before_aneupl_filter, by = "sample")

```

# Sum of allelic counts for all autosomes, per sample

```{r plotting sum for all autosomes (FIGURE), include = TRUE}

all_auto_sum_perSample <- counts_table_by_chr_long[!(chr %in% c("chrM", "chrX", "chrY")), lapply(.SD, sum), .SDcols = c("total", "PWD", "FVB", "unassigned"), by = eval(colnames(design_df_before_aneupl_filter))]
ggplot(all_auto_sum_perSample, aes(x = FVB, y = PWD)) +
  geom_point(aes(color = condition), size = 2, alpha = .7) +
  geom_text_repel(aes(label = sample)) +
  coord_fixed(ratio = 1, xlim = c(10000,400000), ylim = c(10000,400000), expand = TRUE, clip = "on")

```

# Allelic ratio per chromosome

```{r plot per chr, include = TRUE}

# computing FVB/PWD ratio
counts_table_by_chr_long$allelic_ratio_chr <- counts_table_by_chr_long$FVB/(counts_table_by_chr_long$PWD + counts_table_by_chr_long$FVB + 1)

ggplot(counts_table_by_chr_long[!(chr %in% c("chrM", "chrX", "chrY"))], aes(x = chr, y = allelic_ratio_chr, color = condition)) +
  geom_violin(alpha = 0.5) +
  geom_point(aes(group = condition), position = position_jitterdodge(jitter.width = 0.1, dodge.width = 0.9)) +
  geom_text_repel(aes(label = sample), size = 2, max.overlaps = 20, position = position_jitterdodge(jitter.width = 0.1, dodge.width = 0.9)) +
  ggtitle("autosomes")

ggplot(counts_table_by_chr_long[chr == "chrX" & sex == "F"], aes(x = condition, y = allelic_ratio_chr)) +
  geom_violin(alpha = 0.5) +
  geom_point(aes(size = total), position = position_jitter(width = 0.1, seed = 1)) +
  geom_text_repel(aes(label = sample), position = position_jitter(width = 0.1, seed = 1), size = 2, max.overlaps = 30) +
  ggtitle("chrX")

```

* Samples do not lie on the diagonal i.e. PWD counts are in general lower than the FVB ones. This could be due to better annotation of FVB SNPs than PWD ones. 
* No experimental groups with a particularly high number of outliers, which are potential anaeuploidic embryos.
* Outliers (bA10, bA12, bC4, bC19, bD20, bA11, bD5) will be excluded from the rest of the analyses.

```{r sessioninfo, echo = TRUE, results = 'markup'}

sessionInfo()

```



























